compiler/modules/modules.nim

```diff
@@ -53,39 +53,36 @@ template getModuleIdent(graph: ModuleGraph, filename: AbsoluteFile): PIdent =
 
 template packageId(): untyped {.dirty.} = ItemId(module: PackageModuleId, item: int32(fileIdx))
 
-proc getPackage(graph: ModuleGraph; fileIdx: FileIndex): PSym =
-  ## returns package symbol (skPackage) for yet to be defined module for fileIdx
+proc getPackage(graph: ModuleGraph, fileIdx: FileIndex): PSym =
+  ## returns the package symbol (skPackage) for yet to be defined module for
+  ## `fileIdx`
   let
     filename = AbsoluteFile toFullPath(graph.config, fileIdx)
-    name = getModuleIdent(graph, filename)
     info = newLineInfo(fileIdx, 1, 1)
-    pck = getPackageName(graph.config, filename.string)
-    pck2 = if pck.len > 0: pck else: "unknown"
-    pack = getIdent(graph.cache, pck2)
-  
-  result = graph.packageSyms.strTableGet(pack)
-  
-  if result == nil:
-    result = newSym(skPackage, getIdent(graph.cache, pck2), packageId(), nil, info)
-    #initStrTable(packSym.tab)
-    graph.packageSyms.strTableAdd(result)
+    desc = getPkgDesc(graph.config, filename.string)
+    rootPkg = getIdent(graph.cache, desc.pkgRootName)
+    existingRootPkgSym = graph.packageSyms.strTableGet(rootPkg)
+    rootPkgSym =
+      if existingRootPkgSym.isNil:
+        let temp = newSym(skPackage, rootPkg, packageId(), nil, info)
+        graph.packageSyms.strTableAdd(temp)
+        temp
+      else:
+        existingRootPkgSym
+
+  if desc.pkgSubpath == "":
+    result = rootPkgSym
   else:
-    let modules = graph.modulesPerPackage.getOrDefault(result.itemId)
-    let existing = if modules.data.len > 0: strTableGet(modules, name) else: nil
-    if existing != nil and existing.info.fileIndex != info.fileIndex:
-      when false:
-        # we used to produce an error:
-        localReport(graph.config, info,
-          "module names need to be unique per Nimble package; module clashes with " &
-            toFullPath(graph.config, existing.info.fileIndex))
+    let
+      subPkg = getIdent(graph.cache, desc.pkgName)
+      existingSubPkgSym = graph.packageSyms.strTableGet(subPkg)
+    result =
+      if existingSubPkgSym.isNil:
+        let subPkgSym = newSym(skPackage, subPkg, packageId(), rootPkgSym, info)
```

Comment:
Apologies if this is obvious, but what are sub-package symbols needed for? Phrased differently, what would be the implications of also only having the `rootPkgSym` in the `subpath != ""` case?

Reply:
The modules per package, but that's not written to anymore, hmm... that might be a bug that I'm not doing that. The issue is that if we go `foopkg/barmodule`, then it's ambiguous which `barmodule` we actually mean, in case there is more than one:

```
/foopkg
    /a
        barmodule
    /b
        barmodule
```

Reply:
Hm, but providing the full sub-path is always required, no? What I mean is, even if only `foopkg/a/barmodule.nim` would exist, a `foopkg/barmodule` import would still not refer to a file (if I'm understanding the file lookup code correctly, that is).

Reply:
> Apologies if this is obvious, but what are sub-package symbols needed for?

It's not, after having re-reviewed things a bit, the impacts are subtle and mostly correctness oriented. I think the bigger deal is that we have a shot of building on top of the newer approach where we previously didn't.

> Phrased differently, what would be the implications of also only having the `rootPkgSym` in the `subpath != ""` case?

I think the code would truck along and we'd get a few oddities here and there. In particular, we'd have less unique input for things like `sighashes` and some potential vm lookup issues via `vmaux.matches`. These things would likely lie dormant for a long time/not be a problem. But I don't think we could evolve things from that as a base.

In particular the `compilepreter`, which I'd like to ensure isn't too polluted with legacy silliness. I was writing up some integration and I couldn't make sense of package symbols enough such that I felt I'd be able to convert back to `PNode` output for integrating existing backends. I'd have to either implement something from scratch (error prone/second system) or ... not change this up so I can make some sense of it.

Reply:
Thank you for the explanations! Yeah, having the sub-package symbols sounds like the better choice.